spatialprofilingtoolbox/db/feature_matrix_extractor.py

```diff
@@ -96,163 +131,159 @@ def _extract(self,
+                    dataframe[[f'C {m}' for m in criteria.positive_markers]].all(axis=1) &
+                    ~dataframe[[f'C {m}' for m in criteria.negative_markers]].any(axis=1)
+                ).astype(int)
+            matrices[specimen] = MatrixBundle(dataframe, f'{j}.tsv')
+
+        if 'continuous data arrays by specimen' in study:
+            specimens = list(study['continuous data arrays by specimen'].keys())
+            for j, specimen in enumerate(sorted(specimens)):
                 logger.debug('Specimen %s .', specimen)
-                expressions = study['data arrays by specimen'][specimen]
-                number_channels = len(study['target index lookup'])
-                rows = [
-                    self._create_feature_matrix_row(
-                        centroid_coordinates[study_name][specimen][i],
-                        expressions[i],
-                        number_channels,
-                    )
-                    for i in range(len(expressions))
-                ]
+                expression_vectors = study['continuous data arrays by specimen'][specimen]
                 dataframe = DataFrame(
-                    rows,
-                    columns=['pixel x', 'pixel y'] + [f'F{i}' for i in range(number_channels)],
+                    expression_vectors,
+                    columns=[f'C {cs}' for cs in channel_information],
                 )
-                matrices[study_name][specimen] = {
-                    'dataframe': dataframe,
-                    'filename': f'{k}.{j}.tsv',
-                }
-
-            if 'continuous data arrays by specimen' in study:
-                specimens = list(study['continuous data arrays by specimen'].keys())
-                for j, specimen in enumerate(sorted(specimens)):
-                    logger.debug('Specimen %s .', specimen)
-                    expression_vectors = study['continuous data arrays by specimen'][specimen]
-                    number_channels = len(study['target index lookup'])
-                    dataframe = DataFrame(
-                        expression_vectors,
-                        columns=[f'F{i}' for i in range(number_channels)],
-                    )
-                    matrices[study_name][specimen]['continuous dataframe'] = dataframe
+                matrices[specimen].continuous_dataframe = dataframe
 
         logger.info('Done creating feature matrices.')
         return matrices
 
     @staticmethod
-    def _create_feature_matrix_row(centroid, binary, number_channels):
+    def _create_feature_matrix_row(
+        centroid: tuple[float, float],
+        binary: list[str],
+        number_channels: int,
+    ) -> list[float | int]:
         template = '{0:0%sb}' % number_channels   # pylint: disable=consider-using-f-string
-        feature_vector = [int(value) for value in list(template.format(binary)[::-1])]
+        feature_vector: list[int] = [int(value) for value in list(template.format(binary)[::-1])]
```

Comment:
This is the line we would change if we wanted to turn this from ints to bools (not including all the downstream changes we'd need to do...)